File: crates/superposition_provider/examples/local_file_watch_example.rs
use std::path::PathBuf;

use open_feature::EvaluationContext;
use superposition_provider::{
    data_source::file::FileDataSource, local_provider::LocalResolutionProvider,
    traits::AllFeatureProvider, RefreshStrategy, WatchStrategy,
};

#[tokio::main]
async fn main() {
    env_logger::init();

    let manifest_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let config_path = manifest_dir.join("examples/config.toml");

    println!("Watching config file: {:?}", config_path);
    println!("Edit the file in another terminal to see changes.\n");

    let file_source = FileDataSource::new(config_path);

    let provider = LocalResolutionProvider::new(
        Box::new(file_source),
        None,
        RefreshStrategy::Watch(WatchStrategy::default()),
    );
    provider.init().await.unwrap();

    let context = EvaluationContext::default()
        .with_custom_field("os", "linux")
        .with_custom_field("city", "Boston");

    // Poll in a loop to show updated values after file changes
    loop {
        let config = provider.resolve_all_features(&context).await.unwrap();
        println!("Config: {:?}", config);
        tokio::time::sleep(tokio::time::Duration::from_secs(2)).await;
    }
}

File: clients/python/provider/examples/local_file_watch_example.py
"""
Local File Watch Example - Watch a local config file for changes.

Demonstrates file watching capability with automatic reload when the file changes.
Edit the config file in another terminal to see the changes reflected in the running example.
"""

import asyncio
import logging
from pathlib import Path

from openfeature.evaluation_context import EvaluationContext

from superposition_provider import LocalResolutionProvider, FileDataSource
from superposition_provider.types import WatchStrategy

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


async def main():
    """Run the file watch example."""
    # Get the path to the example config file
    example_dir = Path(__file__).parent
    config_path = example_dir / "config.toml"

    logger.info(f"Watching config file: {config_path}")
    logger.info("Edit the file in another terminal to see changes.\n")

    # Create a provider using FileDataSource with WatchStrategy
    provider = LocalResolutionProvider(
        primary_source=FileDataSource(config_path.__str__()),
        refresh_strategy=WatchStrategy(debounce_ms=1000),
    )

    # Initialize the provider
    await provider.initialize(EvaluationContext())

    # Create an evaluation context
    context = EvaluationContext(
        targeting_key=None,
        attributes={
            "os": "linux",
            "city": "Boston",
        },
    )

    logger.info("Polling config every 2 seconds. Press Ctrl-C to stop.\n")

    try:
        # Poll in a loop to show updated values after file changes
        while True:
            config = provider.resolve_all_features(context)
            logger.info(f"Config: {config}")
            await asyncio.sleep(2)
    except KeyboardInterrupt:
        logger.info("\nShutting down...")
    finally:
        await provider.shutdown()
        logger.info("Provider shutdown complete")


if __name__ == "__main__":
    asyncio.run(main())
